components/notebook-controller/controllers/notebook_controller.go

```diff
@@ -567,7 +568,142 @@ func generateVirtualService(instance *v1beta1.Notebook) (*unstructured.Unstructu
 	}
 
 	return vsvc, nil
+}
+
+func generateHTTPRoute(instance *v1beta1.Notebook) (*unstructured.Unstructured, error) {
+	name := instance.Name
+	namespace := instance.Namespace
+	prefix := fmt.Sprintf("/notebook/%s/%s", namespace, name)
+
+	// unpack annotations from Notebook resource
+	annotations := make(map[string]string)
+	for k, v := range instance.ObjectMeta.Annotations {
+		annotations[k] = v
+	}
+
+	pathRewrite := fmt.Sprintf("/notebook/%s/%s", namespace, name)
+	// If AnnotationRewriteURI is present, use this value for path rewrite
+	if _, ok := annotations[AnnotationRewriteURI]; ok && len(annotations[AnnotationRewriteURI]) > 0 {
+		pathRewrite = annotations[AnnotationRewriteURI]
+	}
+
+	httpRoute := &unstructured.Unstructured{}
+	httpRoute.SetAPIVersion("gateway.networking.k8s.io/v1")
+	httpRoute.SetKind("HTTPRoute")
+	httpRoute.SetName(virtualServiceName(name, namespace))
+	httpRoute.SetNamespace(namespace)
+
+	// Get gateway configuration
+	gatewayName := os.Getenv("GATEWAY_NAME")
+	if len(gatewayName) == 0 {
+		gatewayName = "kubeflow-gateway"
+	}
+	gatewayNamespace := os.Getenv("GATEWAY_NAMESPACE")
+	if len(gatewayNamespace) == 0 {
+		gatewayNamespace = "kubeflow"
+	}
+
+	// Set parentRefs for HTTPRoute
+	parentRefs := []interface{}{
+		map[string]interface{}{
+			"name":      gatewayName,
+			"namespace": gatewayNamespace,
+		},
+	}
+	if err := unstructured.SetNestedSlice(httpRoute.Object, parentRefs, "spec", "parentRefs"); err != nil {
+		return nil, fmt.Errorf("set .spec.parentRefs error: %v", err)
+	}
+
```

Comment:
I'm not sure where this is needed but I 'm not too familiar with error handling in Go so I 'll let someone else comment on this one